From e00dbf43d739cbd327034cc5044de4eb4c0e4e78 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=A0=D1=83=D1=81=D0=BB=D0=B0=D0=BD=20=D0=98=D0=B6=D0=B1?=
 =?UTF-8?q?=D1=83=D0=BB=D0=B0=D1=82=D0=BE=D0=B2?= <lrn1986@gmail.com>
Date: Wed, 2 Apr 2014 15:47:28 +0000
Subject: [PATCH] W32: Allow shared libs to be built

This requires -no-undefined flag and resolving all dependencies at link-time by
linking plugins to libpeas and libpeas-gtk as needed.

https://bugzilla.gnome.org/show_bug.cgi?id=727499
---
 configure.ac                                               | 4 ++--
 loaders/python/Makefile.am                                 | 5 +++--
 loaders/python3/Makefile.am                                | 5 +++--
 loaders/seed/Makefile.am                                   | 5 ++++-
 peas-demo/plugins/helloworld/Makefile.am                   | 6 +++++-
 peas-demo/plugins/secondtime/Makefile.am                   | 5 ++++-
 tests/libpeas-gtk/plugins/builtin-configurable/Makefile.am | 6 +++++-
 tests/libpeas-gtk/plugins/configurable/Makefile.am         | 6 +++++-
 tests/plugins/builtin/Makefile.am                          | 4 +++-
 tests/plugins/has-dep/Makefile.am                          | 4 +++-
 tests/plugins/loadable/Makefile.am                         | 4 +++-
 tests/plugins/self-dep/Makefile.am                         | 4 +++-
 12 files changed, 43 insertions(+), 15 deletions(-)

diff --git a/configure.ac b/configure.ac
index 89e46df..e6e5fa3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -395,12 +395,12 @@ AC_SUBST(LT_AGE)
 # The reasons for -rpath /nowhere can be found at:
 # http://www.mail-archive.com/automake@gnu.org/msg10725.htm
 
-PLUGIN_LIBTOOL_FLAGS="-module -avoid-version"
+PLUGIN_LIBTOOL_FLAGS="-module -avoid-version -no-undefined"
 TEST_PLUGIN_LIBTOOL_FLAGS="$PLUGIN_LIBTOOL_FLAGS -rpath /nowhere"
 TEST_INTROSPECTION_LIBTOOL_FLAGS="-rpath /nowhere -export-dynamic -no-undefined"
 # Cannot use "-export-symbols-regex peas_register_types"
 # because it might make the python plugin loader not work!
-LOADER_LIBTOOL_FLAGS="-module -avoid-version"
+LOADER_LIBTOOL_FLAGS="-module -avoid-version -no-undefined"
 AC_SUBST(PLUGIN_LIBTOOL_FLAGS)
 AC_SUBST(TEST_PLUGIN_LIBTOOL_FLAGS)
 AC_SUBST(TEST_INTROSPECTION_LIBTOOL_FLAGS)
diff --git a/loaders/python/Makefile.am b/loaders/python/Makefile.am
index a33d0af..38cc98c 100644
--- a/loaders/python/Makefile.am
+++ b/loaders/python/Makefile.am
@@ -25,8 +25,9 @@ libpythonloader_la_LDFLAGS = \
 	$(PYTHON2_LDFLAGS)
 
 libpythonloader_la_LIBADD = \
-	$(PEAS_LIBS)			\
-	$(PYGOBJECT_LIBS)		\
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)				\
+	$(PYGOBJECT_LIBS)			\
 	$(PYTHON2_LIBS)
 
 gcov_sources = $(libpythonloader_la_SOURCES)
diff --git a/loaders/python3/Makefile.am b/loaders/python3/Makefile.am
index 884d710..19e085e 100644
--- a/loaders/python3/Makefile.am
+++ b/loaders/python3/Makefile.am
@@ -25,8 +25,9 @@ libpython3loader_la_LDFLAGS = \
 	$(PYTHON3_LDFLAGS)
 
 libpython3loader_la_LIBADD = \
-	$(PEAS_LIBS)			\
-	$(PYGOBJECT_LIBS)		\
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)				\
+	$(PYGOBJECT_LIBS)			\
 	$(PYTHON3_LIBS)
 
 gcov_sources = $(libpython3loader_la_SOURCES)
diff --git a/loaders/seed/Makefile.am b/loaders/seed/Makefile.am
index 75566c0..12ae3b9 100644
--- a/loaders/seed/Makefile.am
+++ b/loaders/seed/Makefile.am
@@ -19,7 +19,10 @@ libseedloader_la_SOURCES = \
 	peas-plugin-loader-seed.h
 
 libseedloader_la_LDFLAGS = $(LOADER_LIBTOOL_FLAGS) $(GCOV_LDFLAGS)
-libseedloader_la_LIBADD = $(PEAS_LIBS) $(SEED_LIBS)
+libseedloader_la_LIBADD = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS) 				\
+	$(SEED_LIBS)
 
 gcov_sources = $(libseedloader_la_SOURCES)
 include $(top_srcdir)/Makefile.gcov
diff --git a/peas-demo/plugins/helloworld/Makefile.am b/peas-demo/plugins/helloworld/Makefile.am
index a7b151c..7519d26 100644
--- a/peas-demo/plugins/helloworld/Makefile.am
+++ b/peas-demo/plugins/helloworld/Makefile.am
@@ -14,7 +14,11 @@ libhelloworld_la_SOURCES = \
 	peasdemo-hello-world-configurable.c
 
 libhelloworld_la_LDFLAGS = $(PLUGIN_LIBTOOL_FLAGS)
-libhelloworld_la_LIBADD  = $(PEAS_LIBS) $(PEAS_GTK_LIBS)
+libhelloworld_la_LIBADD  = \
+	$(top_builddir)/libpeas-gtk/libpeas-gtk-1.0.la	\
+	$(top_builddir)/libpeas/libpeas-1.0.la		\
+	$(PEAS_LIBS)					\
+	$(PEAS_GTK_LIBS)
 
 plugin_DATA = helloworld.plugin
 
diff --git a/peas-demo/plugins/secondtime/Makefile.am b/peas-demo/plugins/secondtime/Makefile.am
index a20e76c..e54e5d3 100644
--- a/peas-demo/plugins/secondtime/Makefile.am
+++ b/peas-demo/plugins/secondtime/Makefile.am
@@ -12,7 +12,10 @@ libsecondtime_la_SOURCES = \
 	second-time.c
 
 libsecondtime_la_LDFLAGS = $(PLUGIN_LIBTOOL_FLAGS)
-libsecondtime_la_LIBADD  = $(PEAS_LIBS) $(PEAS_GTK_LIBS)
+libsecondtime_la_LIBADD  = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS) 				\
+	$(PEAS_GTK_LIBS)
 
 plugin_DATA = secondtime.plugin
 
diff --git a/tests/libpeas-gtk/plugins/builtin-configurable/Makefile.am b/tests/libpeas-gtk/plugins/builtin-configurable/Makefile.am
index 78d9529..53577d6 100644
--- a/tests/libpeas-gtk/plugins/builtin-configurable/Makefile.am
+++ b/tests/libpeas-gtk/plugins/builtin-configurable/Makefile.am
@@ -16,7 +16,11 @@ libbuiltin_configurable_la_SOURCES = \
 	builtin-configurable.h
 
 libbuiltin_configurable_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libbuiltin_configurable_la_LIBADD  = $(PEAS_LIBS) $(PEAS_GTK_LIBS)
+libbuiltin_configurable_la_LIBADD  = \
+	$(top_builddir)/libpeas-gtk/libpeas-gtk-1.0.la	\
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS) 				\
+	$(PEAS_GTK_LIBS)
 
 noinst_PLUGIN = builtin-configurable.plugin
 
diff --git a/tests/libpeas-gtk/plugins/configurable/Makefile.am b/tests/libpeas-gtk/plugins/configurable/Makefile.am
index 7588069..77536cc 100644
--- a/tests/libpeas-gtk/plugins/configurable/Makefile.am
+++ b/tests/libpeas-gtk/plugins/configurable/Makefile.am
@@ -14,7 +14,11 @@ libconfigurable_la_SOURCES = \
 	configurable-plugin.h
 
 libconfigurable_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libconfigurable_la_LIBADD  = $(PEAS_LIBS) $(PEAS_GTK_LIBS)
+libconfigurable_la_LIBADD  = \
+	$(top_builddir)/libpeas-gtk/libpeas-gtk-1.0.la	\
+	$(top_builddir)/libpeas/libpeas-1.0.la		\
+	$(PEAS_LIBS) 					\
+	$(PEAS_GTK_LIBS)
 
 noinst_PLUGIN = configurable.plugin
 
diff --git a/tests/plugins/builtin/Makefile.am b/tests/plugins/builtin/Makefile.am
index b909ef3..8d01fa8 100644
--- a/tests/plugins/builtin/Makefile.am
+++ b/tests/plugins/builtin/Makefile.am
@@ -13,7 +13,9 @@ libbuiltin_la_SOURCES = \
 	builtin-plugin.h
 
 libbuiltin_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libbuiltin_la_LIBADD  = $(PEAS_LIBS)
+libbuiltin_la_LIBADD  = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)
 
 noinst_PLUGIN = builtin.plugin
 
diff --git a/tests/plugins/has-dep/Makefile.am b/tests/plugins/has-dep/Makefile.am
index facb48d..3c6dfd4 100644
--- a/tests/plugins/has-dep/Makefile.am
+++ b/tests/plugins/has-dep/Makefile.am
@@ -13,7 +13,9 @@ libhas_dep_la_SOURCES = \
 	has-dep-plugin.h
 
 libhas_dep_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libhas_dep_la_LIBADD  = $(PEAS_LIBS)
+libhas_dep_la_LIBADD  = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)
 
 noinst_PLUGIN = has-dep.plugin
 
diff --git a/tests/plugins/loadable/Makefile.am b/tests/plugins/loadable/Makefile.am
index c387e5e..46bb90c 100644
--- a/tests/plugins/loadable/Makefile.am
+++ b/tests/plugins/loadable/Makefile.am
@@ -13,7 +13,9 @@ libloadable_la_SOURCES = \
 	loadable-plugin.h
 
 libloadable_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libloadable_la_LIBADD  = $(PEAS_LIBS)
+libloadable_la_LIBADD  = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)
 
 noinst_PLUGIN = loadable.plugin
 
diff --git a/tests/plugins/self-dep/Makefile.am b/tests/plugins/self-dep/Makefile.am
index f20079c..ff5ce90 100644
--- a/tests/plugins/self-dep/Makefile.am
+++ b/tests/plugins/self-dep/Makefile.am
@@ -13,7 +13,9 @@ libself_dep_la_SOURCES = \
 	self-dep-plugin.h
 
 libself_dep_la_LDFLAGS = $(TEST_PLUGIN_LIBTOOL_FLAGS)
-libself_dep_la_LIBADD  = $(PEAS_LIBS)
+libself_dep_la_LIBADD  = \
+	$(top_builddir)/libpeas/libpeas-1.0.la	\
+	$(PEAS_LIBS)
 
 noinst_PLUGIN = self-dep.plugin
 
-- 
1.8.5.3
